### PR TITLE
Add copy file endpoint code

### PR DESCRIPTION
**Pull request by sathuhebbar**
Added file `src/endpoints/file.copyFile.ts`, modified `src/app.ts`:

1. Addresses #26 
2. Tests have not yet been implemented
3. Error messages and handling are still rudimentary, will work on them

### CODE REVIEW

File: src/endpoints/file/copyFile.ts [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    let overwrite = false;
+
+    if (query.overwrite) {
+      if (query.overwrite === "true")
+        overwrite = true;
+      else if (query.overwrite === "false")
+        overwrite = false;
+      else {
+        set.status = 400;
+        return 'Overwrite must be one of "true" and "false"';
```
Comment (Xkonti) [inline src/endpoints/file/copyFile.ts]:
```
Parameter "overwrite" must be either "true" or "false"
```

File: src/endpoints/file/copyFile.ts [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    relativeSourcePath = relativeSourcePath as string;
+
+    relativeDestinationPath = relativeDestinationPath as string;
+
+    // Get the full path to the SOURCE file
+    const sourceFilePath = join(getConfig().dataDir, relativeSourcePath);
+
+    // Get the full path to the DESTINATION file
+    const destinationFilePath = join(getConfig().dataDir, relativeDestinationPath);
```
Comment (Xkonti) [inline src/endpoints/file/copyFile.ts]: Look at `validateRelativePath`:

https://github.com/Xkonti/file-api/blob/bd3b6fbfd75e3aea70dae70f8f9d1470f91a2869/src/utils/pathUtils.ts#L19

It returns both relative and absolute paths. It's usage looks like this:
https://github.com/Xkonti/file-api/blob/bd3b6fbfd75e3aea70dae70f8f9d1470f91a2869/src/endpoints/file/uploadFile.ts#L25-L31

File: src/endpoints/file/copyFile.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+    var file = null;
```
Comment (Xkonti) [inline src/endpoints/file/copyFile.ts]: Never `var`. Use `let` instead.
Also, let's rename it to `sourceFile` to be more clear.

File: src/endpoints/file/copyFile.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    else {
+      if (fileWriteResult.error == fileAlreadyExistsMsg) {
+        set.status = 409;
+        return "File exists and overwrite is not set";
+      }
+      else {
+        set.status = 500;
+        return `An error occurred while copying the file ${fileWriteResult.error}`;
+      }
+    }
```
Comment (Xkonti) [inline src/endpoints/file/copyFile.ts]: Maybe we could extract this functionality in a form of `copyFile` that would live inside the `fileUtils.ts`? Other endpoints will reuse that functionality.

**Review by Xkonti — CHANGES_REQUESTED** [review]
Nice so far!

File: src/endpoints/file/copyFile.ts (hunk 1 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Right. Made this message a constant in `commonResponses.ts` and am using that.

File: src/endpoints/file/copyFile.ts (hunk 3 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Right, fixed.

File: src/endpoints/file/copyFile.ts (hunk 2 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Right, modified code to use this instead.

File: src/endpoints/file/copyFile.ts (hunk 4 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Yup, moved this functionality to `fileUtils.ts`.

**Comment by sathuhebbar** [comment]
Added these changes, additionally modified `src/constants/commonResponses.ts` to add constants for the new error messages.
Please have a look.

File: src/endpoints/file/copyFile.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+} from '../../constants/commonResponses'
+
+export function addCopyFileEndpoint(app: Elysia) {
+  return app.post('file/copy', async ({ query, set }) => {
+
+    // Verify that the SOURCE file path is valid
+    let relativeSourcePath = query.source ? (query.source as string) : "";
+
+    // Get the DESTINATION file path
+    let relativeDestinationPath = query.destination ? (query.destination as string) : "";
```
Comment (Xkonti) [inline src/endpoints/file/copyFile.ts]: You can replace these with the `validateRelativePath` checks from below. From the result you can get both relative and absolute paths:

```
const sourcePath = sourcePathValidationResult.value; // assignment to quick access
let something = sourcePath.relativePath; // example of accessing it
```

File: src/endpoints/file/copyFile.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+      return invalidSourcePathMsg;
+    }
+    const absoluteSourcePath = sourcePathValidationResult.value.absolutePath;
+
+    const destinationPathValidationResult = validateRelativePath(relativeDestinationPath);
+    if (destinationPathValidationResult.isErr()) {
+      set.status = 400;
+      return invalidDestinationPathMsg;
+    }
+    const absoluteDestinationPath = destinationPathValidationResult.value.absolutePath;
```
Comment (Xkonti) [inline src/endpoints/file/copyFile.ts]: This should go to the beginning to replace initial verification (as per previous comment)

File: .gitignore [hunk 7]
```diff
@@ -40,4 +40,9 @@ yarn-error.log*
...
 package-lock.json
 **/*.bun
-idea/
\ No newline at end of file
+idea/
+
+# utility script to rebuild docker
+# image and restart the API after
+# changes to code
+/runner-script.sh
```
Comment (Xkonti) [inline .gitignore]: There's the script mentioned in `.gitignore`, but no script is included in the PR.

**Review by Xkonti — CHANGES_REQUESTED** [review]

File: src/endpoints/file/copyFile.ts (hunk 5 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Right, so I can directly pass `query.source` to `validateRelativePath` and skip this step because it already checks for `null` there. I can also eliminate the `absoluteXPath` variables and access them using `Xpath.absoluteValue`.

Is that right?

File: .gitignore (hunk 7 above)
Comment (sathuhebbar) [reply .gitignore]: This was a script that I used to test the program locally, but I didn't want it included in the commit, so I added it to `.gitignore`. Can you please tell me what I should do in this case?

File: src/endpoints/file/copyFile.ts (hunk 6 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Moved as required.

File: src/endpoints/file/copyFile.ts (hunk 5 above)
Comment (Xkonti) [reply src/endpoints/file/copyFile.ts]: That is correct. You can either access those values via `result.value.absolutePath`/`result.value.relativePath` or destructure them:

```ts
const { absolutePath: absoluteDestinationPath, relativePath: relativeDestinationPath } = result.value;
```

Whichever you deem more suitable.

File: .gitignore (hunk 7 above)
Comment (Xkonti) [reply .gitignore]: In this case you could just remove the lines you added to `.gitinore` (from empty line 44 to line 48) and make sure you don't commit your script file.

File: src/endpoints/file/copyFile.ts (hunk 5 above)
Comment (sathuhebbar) [reply src/endpoints/file/copyFile.ts]: Right, fixed this.

File: .gitignore (hunk 7 above)
Comment (sathuhebbar) [reply .gitignore]: Right, fixed this.

**Comment by sathuhebbar** [comment]
Made the required modifications. Can you please tell me about this failing workflow test? I don't really understand what it means.

Thanks!

**Comment by Xkonti** [comment]
If you click on _details_ next to the failed `Test PR / check-pr (pull_request)` workflow it will display the the log. There at the end you can see that all tests have passed. And right before that you have a table showing the test coverage:

---------------------------------------|---------|---------|-------------------
File                                   | % Funcs | % Lines | Uncovered Line #s
---------------------------------------|---------|---------|-------------------
All files                              |   96.11 |   88.48 |
 src/app.ts                            |  100.00 |  100.00 | 
 src/constants/commonResponses.ts      |  100.00 |  100.00 | 
 src/endpoints/errors.ts               |  100.00 |  100.00 | 
 src/endpoints/file/checkFileExists.ts |  100.00 |   92.00 | 29-30
 **src/endpoints/file/copyFile.ts**    |   **50.00** ⚠️ |    **8.93** ⚠️ | 13-63
 src/endpoints/file/getFile.ts         |  100.00 |   91.67 | 28-29
 src/endpoints/file/getFileSize.ts     |  100.00 |   91.67 | 29-30
 src/endpoints/file/uploadFile.ts      |  100.00 |   77.98 | 67,70-79,81-82,122,125-127,129-130,144,153-154,158
 src/endpoints/guards.ts               |  100.00 |  100.00 | 
 src/endpoints/list/getList.ts         |  100.00 |   86.67 | 28,30-31
 src/testing/constants.ts              |  100.00 |  100.00 | 
 src/testing/testFileSystem.ts         |  100.00 |   94.74 | 24-26
 src/utils/config.ts                   |  100.00 |  100.00 | 
 src/utils/directoryUtils.ts           |  100.00 |   95.06 | 23,71,88
 **src/utils/fileUtils.ts**            |   80.00 |   **53.97** ⚠️ | 55,63,80-106
 src/utils/pathUtils.ts                |  100.00 |  100.00 | 
 src/utils/requestBuilder.ts           |  100.00 |  100.00 | 
 src/utils/urlBuilder.ts               |  100.00 |  100.00 | 
---------------------------------------|---------|---------|-------------------

The project is configured to fail pull requests if any of the files' test coverage is below 70% of functions or lines:
https://github.com/Xkonti/file-api/blob/bd3b6fbfd75e3aea70dae70f8f9d1470f91a2869/bunfig.toml#L1-L3

In the table I highlighted both `copyFile.ts` and `fileUtils.ts` that don't have enough test coverage.

You can look at `uploadFile.ts` and `getFile.ts` endpoints to see how they are tested (different edge cases etc) and then how other methods if `fileUtils.ts` are tested.
